torchaudio/models/wav2vec2/model.py

```diff
@@ -116,6 +116,125 @@ def forward(
+        logits /= logit_temp
+        if neg_is_pos.any():
+            logits[1:][neg_is_pos] = float("-inf")
+        logits = logits.transpose(0, 1)  # (num_x, num_cls+1)
+        return logits
+
+    def _get_padding_mask(self, x: Tensor, lengths: Tensor) -> Tensor:
+        batch_size, max_len, _ = x.shape
+        # create mask for padded elements and zero-out them
+        mask = torch.arange(max_len, device=lengths.device).expand(batch_size, max_len) >= lengths[:, None]
+        return mask
+
+    def forward(
+            self,
+            waveforms: Tensor,
+            labels: Tensor,
+            audio_lengths: Optional[Tensor] = None,
+    ) -> Tuple[Tensor, Optional[Tensor]]:
+        """Compute the sequence of probability distribution over labels.
+
+        Args:
+            waveforms (Tensor): Audio tensor of shape `(batch, frames)`.
+            labels (Tensor): Label for Pre-training or finetuning task.
+            audio_lengths (Tensor or None, optional):
+                Indicates the valid length of each audio in the batch.
+                Shape: `(batch, )`.
+                When the ``waveforms`` contains audios with different durations,
+                by providing ``lengths`` argument, the model will compute
+                the corresponding valid output lengths and apply proper mask in
+                transformer attention layer.
+                If ``None``, it is assumed that all the audio in ``waveforms``
+                have valid length. Default: ``None``.
+
+        Returns:
+            (Tensor, Optional[Tensor]):
+            Tensor
+                The sequences of probability distribution (in logit) with masking.
+                Shape: `(batch, frames, num labels)`.
+            Tensor
+                The sequence of probability distribution (in logit) without masking.
+                Shape: `(batch, frames, num labels)`.
+            Tensor or None
+                If ``label_lengths`` argument was provided, a Tensor of shape `(batch, )`
+                is returned.
+                It indicates the valid length in time axis of the output Tensor.
+        """
+        x, lengths = self.feature_extractor(waveforms, audio_lengths)
+        padding_mask = self._get_padding_mask(x, lengths)
+        x, attention_mask = self.encoder._preprocess(x, lengths)
+        x, mask = self.mask_generator(x, padding_mask, self.mask_embedding)
+
+        x = self.encoder.transformer(x, attention_mask=attention_mask)
+        proj_x = self.final_proj(x)
+        mask_m = torch.logical_and(~padding_mask, mask)
+        mask_u = torch.logical_and(~padding_mask, ~mask_m)
+        proj_x_m = proj_x[mask_m]
+        label_m = labels[mask_m]
+        logit_m = self._compute_pred(proj_x_m, label_m, self.label_embeddings)
+
+        proj_x_u = proj_x[mask_u]
+        label_u = labels[mask_u]
+        logit_u = self._compute_pred(proj_x_u, label_u, self.label_embeddings)
```

Comment:
Looking at the implementation of Wav2Vec2, it seems that it could be used without modifications (or with minimal modifications) to be able to handle HuBERT.

1 - The `_get_padding_mask`, and `mask_generator` can be moved inside `encoder`. This can be achieve either with a wrapper class around the `encoder`, or by extending the original encoder used in `Wav2Vec2`.
2 - the `final_proj` and the logic that follows (including `_compute_pred` can be contained in an new `nn.Module` which will be passed as the `aux` of the model.

Here is some example snippets that illustrates my points:

```python
# as is
class Wav2Vec2(...):
   def __init__(self, feature_extractor, encoder, aux=None):
       ...

# either we extend the original Encoder
# or we wrap it
class EncoderWithMask(...):
    def forward(self, input):
        mask = generate_mask(input)
        out = self.encoder(input, mask)
        # we return the mask, and it's up to you to
        # decide if you want to potentially support that
        # return value in the Wav2Vec2
        return out, mask
# create a new Aux class that will
# be passed to the wav2vec2 model
class MyAux(nn.Module):
    def __init__(self, ...):
        self.final_proj = nn.Linear(...)
        self.label_embedding = nn.Embedding(...)
    
    def forward(self, input):
       x, mask = input
       out = self.final_proj(x)
       ... = self.label_embedding(targets)
       ... = self_compute_pred(...)
        return ...

# instantiate the HuBERT models
def hubert_pretrain_model(...):
    return Wav2Vec2(feature_extractor, EncoderWithMask(...), aux=MyAux(...))



```

Reply:
@fmassa Thanks for the great input! The design looks reasonable, as the HuBERT model can be initialized the same way as Wav2Vec2. 
One issue is that for pre-training, the ``labels`` is necessary in the forward method, so we can't reuse the same forward method in Wav2Vec2. I was thinking about moving the label-related computation to loss function, however, there are some trainable parameters such as ``mask_embedding`` and ``label_embeddings``, if moving them to loss function, how to add them to the optimizer is tricky then.